### ISSUE TITLE
chapter 5: 03_pyfolio_demo.ipynb

### ISSUE THREAD

**Issue by Eeonum** [stefan-jansen/machine-learning-for-trading#113]
I am getting the following error running the above code:

ValueError: unsupported pickle protocol: 5

**Comment by stefan-jansen** [comment]
Are you running both the notebook that creates the file as well as the one that reads it using the backtest environment of the new docker image?

**Comment by Eeonum** [comment]
Thanks Stefan. The issue is the consistency.

Here is what I did:

1. I am using my own env python 3.6 with the latest version of zipline
2. I created assets.h5 using the same env to be consistent
3. everything is working fine now.

**Comment by Eeonum** [comment]
The new error is this:

` File "/Users/jennahaerr/Dropbox (Life Elixir)/Companies/Eonum/gitHub_v3/machine-learning-for-trading-eonum_v1/05_strategy_evaluation/03_pyfolio_demo.py", line 255, in <module>
    (interesting_times['Fall2015']
KeyError: 'Fall2015'`

**Comment by stefan-jansen** [comment]
I just reran the notebooks `backtest_with_trades` and `pyfolio_demo`using the current Docker image (with PyFolio 0.9.2) and cannot reproduce any of these errors. Are you using the recommended environment(s)?

**Comment by Eeonum** [comment]
You are using a docker file which is called ml4t-zipline. Note that zipline is under python3.5 which is depreciated. I am using the my own environment under python 3.6 with the latest zipline. I wonder why this error comes up.

**Comment by stefan-jansen** [comment]
The current Docker images uses Python 3.6 (with a patched version of [Zipline 1.4.1](https://github.com/stefan-jansen/zipline)) and 3.8 (for everything but Zipline-related libraries). 

It only contains two environments, `ml4t` and `backtest` as described [here](https://github.com/stefan-jansen/machine-learning-for-trading/tree/master/installation). I would recommend using this Docker image.

**Comment by life-elixir** [comment]
Thank you. When I run the command 
`docker run -it -v $(pwd):/home/packt/ml4t -p 8888:8888 -e QUANDL_API_KEY=<APIkey> --name ml4t appliedai/packt:latest bash`
it takes me to the old docker files with ml4t and ml4t-zipline environments. Should I redownload the docker images to have a new one? If yes, what would be the command to download the version 2 image?

**Comment by stefan-jansen** [comment]
You can use ```docker pull appliedai/packt:latest``` to obtain the latest version. I've updated the instructions accordingly.

**Comment by life-elixir** [comment]
Thank you. Why the new container doesn't have the ml4t-dl for deep learning?

**Comment by stefan-jansen** [comment]
I've tried to simplify things (and reduce the image size) so you only need 2 envs. `ml4t` is intended for all notebooks except those that need Zipline.

**Comment by life-elixir** [comment]
For biplane purposes, I had to go back to the example chapter4/04_single_factor_zipline.ipynb using backtest env in the latest docker/container generates the following error. Also, it is better to automatically assign the proper env to each Jupiter notebook because I had to assign it to the backtest env myself. Thank you!

`TypeError: _run() missing 1 required positional argument: 'custom_loader'`

**Comment by stefan-jansen** [comment]
Thanks, I had indeed forgotten to test the command line version, sorry about that. I've updated Zipline as well as the notebook; the new Docker image should be on Docker Hub in a few minutes.

**Comment by life-elixir** [comment]
Last time I checked, I couldn't find the bundles in the data folder after ingestion. I will see if that one is still an issue. Thanks Stefan.

**Comment by stefan-jansen** [comment]
There should be a (hidden) `.zipline` subdirectory in your data folder because the docker file creates the environment variable
```
ENV ZIPLINE_ROOT=$HOME/ml4t/data/.zipline
```
This would however not work as intended in case you do not launch the container from the root directory of the cloned repo.

**Comment by life-elixir** [comment]
chapter4/04_single_factor_zipline.ipynb using backtest env in the latest docker/container generates the following error again. 

`chapter4/04_single_factor_zipline.ipynb using backtest env in the latest docker/container generates the following error. `

**Comment by stefan-jansen** [comment]
I just downloaded the latest image and ran the file one more time without this error which was due to the invocation of `run_algorithm()` via the Zipline jupyter notebook extension; the other Zipline notebooks call this function directly. The [latest commit](https://github.com/stefan-jansen/zipline/commit/3a431ba13ad6bcaad1186b2efe21991a636628dc) fixes this and the Docker images uses this version for the `backtest` environment. 

Since my test just succeeded, were you perhaps still using the previous version of the image that did indeed produce this error?
